2017-1/snRNA/ex7/readme.md

```diff
@@ -0,0 +1,38 @@
+# 实验七  ansible 一键部署 dhcp nfs ftp samba dns 服务  
+ 
+
+###  实验要求：  
+使用ansible技术重构FTP、NFS、DHCP、DNS、Samba服务器的自动安装与自动配置
+
+
+### 实验过程：
+
+首先参考[课件上的内容](https://sec.cuc.edu.cn/huangwei/course/LinuxSysAdmin/chap0x08.md.html#/12/2)进行ansible部署的前期准备工作
+
+编写的ansible脚本的思路与之前的bash脚本的一键部署相类似。
+
+- 在group_vars中的all文件中声明配置的变量的信息  
+ 
+- roles中每个目录分别承担各自服务的安装与配置  
+ 
+- 每个roles中的tasks负责执行主要的安装配置功能  
+
+- templates文件夹中放置各个服务的配置文件的模板
+  
+- handles 中进行服务的重启  
+  
+
+![](image/1.png)    
+![](image/2.png)  
+ 
+### 实验遇到的问题  
+ 
+1. 部署服务时需要进行例如ftp 、smb 用户的添加等步骤，存在需要输入密码的情况。
+
+   - 采用expect模块，但是一直存在问题 str object  has no attribute of decode 

```

Comment:
不需要用expect模块，我推荐的实现方法有2种：

1. 根据明文password生成对应的hashed password。在ansible脚本或目标passwd配置文件（例如：/etc/proftpd/ftp.passwd ）里硬编码hashed password（按照目标配置文件的格式预先生成好）。Linux系统种支持的hashed password算法和格式我们在上一个夏季小学期时讲过，你们也用过。具体如FTP、SMB等不同服务支持和启用的hashed password需要查看对应的帮助手册。普遍支持的hash算法和格式可以`man 3 crypt`。
2. 借助ansible的[Vault](http://docs.ansible.com/ansible/playbooks_vault.html) 模块，将明文的ansible配置脚本进行对称加密，每次执行ansible脚本的时候在命令行输入一次passpharse解密后执行。对于ansible 2.3+版本，还可以在明文文件中引用独立的加密变量，这样可以避免对整个脚本加密，而只需要对特定需要加密保护的变量进行加密即可。

所谓交互式输入口令都可以通过相关密码管理程序的非交互式输入命令行参数来变成无人值守模式，例如`ftpasswd`的参数：

```
--stdin
                Read the password directly from standard in rather than
                prompting for it.  This is useful for writing scripts that
                automate use of ftpasswd.
```